### PR TITLE
#0: Fix compilation errors when using g++-11 and fmt v10

### PR DESCRIPTION
**Pull request by razorback3**
I tried building tt-metal project with g++-11 (with C++20 standard) and encountered some compilation errors (initially warnings but turned into errors because of -Werror flag).

This PR fixes some compilation errors when using g++-11 and fmt v10 (it is v8 in the current metal repo but I guess we will move to higher version someday 😄).

The compilation errors this PR solves:
- Maybe uninitialized
- Narrowing conversion
- Missing header inclusion
- Static variable not permitted in a constexpr function
- Bypass boost-ext/reflect (C++20 minimal static reflection library) unittest for g++-11 (it requires g++-12 or higher)
- Check format string at runtime instead of compile-time ([more info](https://fmt.dev/latest/api.html#compile-time-format-string-checks))

The main reason I am trying with g++-11 is that I am building an external project that links metal libraries ([libtt_metal.so](http://libtt_metal.so/) and [libtt_eager.so](http://libtt_eager.so/)). The external project is built upon g++-11.

Post commit test: https://github.com/tenstorrent/tt-metal/actions/runs/9646947558

### CODE REVIEW

**Comment by razorback3** [comment]
@TT-billteng 
May I ask you who I have to ask for review for this change to be approved?
Code owners are already added to the reviewers.

**Review by bbradelTT — APPROVED** [review]

File: tt_metal/common/logger.hpp
Comment (ayerofieiev-tt) [inline tt_metal/common/logger.hpp:126]: Can you clarify what changed between versions?

File: ttnn/cpp/ttnn/device_operation.hpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/device_operation.hpp:92]: @arakhmati ptal

File: ttnn/cpp/ttnn/operations/reduction/argmax/device/argmax_program_factory.hpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/reduction/argmax/device/argmax_program_factory.hpp:84]: Let’s avoid c-style cast, use static_cast instead. Thank you!

**Review by ayerofieiev-tt — APPROVED** [review]
Thank you! Looks good to me. @TT-billteng can we have a job which will tries compiling the codebase with gcc toolset?

Please wait for @arakhmati to check the variant mapping portion.

File: tt_metal/common/logger.hpp
Comment (razorback3) [reply tt_metal/common/logger.hpp:126]: https://fmt.dev/latest/api.html#compile-time-format-string-checks
In the latest fmt library (v10), compile-time format string checks are enabled by default on compilers that support C++20.
Since _fmt_ variable in the function is _char const*_ type, it cannot be checked in compile-time thus we should wrap it with _fmt::runtime_.
https://github.com/tenstorrent/tt-metal/blob/8289f9fe531cc4b3f6bcc17cbbc8f992b03f2a63/tt_metal/common/logger.hpp#L109

File: ttnn/cpp/ttnn/operations/reduction/argmax/device/argmax_program_factory.hpp
Comment (razorback3) [reply ttnn/cpp/ttnn/operations/reduction/argmax/device/argmax_program_factory.hpp:84]: OK. I followed the style in the original code.
You mean changing all c-style cast in the file, right?

Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/reduction/argmax/device/argmax_program_factory.hpp:84]: Oh, alright. You did the right thing then. Thank you for the clarification

File: tt_metal/common/logger.hpp
Comment (ayerofieiev-tt) [reply tt_metal/common/logger.hpp:126]: Oh, good to know.. Thank you.

**Comment by TT-billteng** [comment]
> Thank you! Looks good to me. @TT-billteng can we have a job which will tries compiling the codebase with gcc toolset?
> 
> Please wait for @arakhmati to check the variant mapping portion.

Could try it after we have 22.04/24.04 docker working with Metal

File: ttnn/cpp/ttnn/device_operation.hpp
Comment (razorback3) [reply ttnn/cpp/ttnn/device_operation.hpp:92]: Any comment from @arakhmati ? 🙏

Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/device_operation.hpp:92]: He got sick, should be back tomorrow.

**Review by arakhmati — APPROVED** [review]

File: ttnn/cpp/ttnn/device_operation.hpp
Comment (arakhmati) [reply ttnn/cpp/ttnn/device_operation.hpp:92]: Looks okay to me. I believe the compiler should do the same thing whether it's contstexpr or not

**Review by TT-BrianLiu — APPROVED** [review]

**Comment by razorback3** [comment]
> > Thank you! Looks good to me. @TT-billteng can we have a job which will tries compiling the codebase with gcc toolset?
> > Please wait for @arakhmati to check the variant mapping portion.
> 
> Could try it after we have 22.04/24.04 docker working with Metal

Is there any schedule when Metal support 22.04+?

**Comment by razorback3** [comment]
1. As Metal moved to use clang and libc++ library with C++20 standard, it killed our external project because ours were all libstdc++ with C++17 standard based.
2. Therefore, we moved our external project to g++-11 which support C++20 standard. Also, we built Metal with g++-11 and libstdc++. It worked for a short period of time before [this](https://github.com/tenstorrent/tt-metal/commit/a8e9e22476c96d60b86f284f2f4eb83760805968) and [this](https://github.com/tenstorrent/tt-metal/commit/b5e0888819eb18fa15a8f639532bfcf662276b5d).

3. The above two commits require g++-12 or above because the reflect library those commits are using does not support g++-11. To get g++-12 or above, we should use Ubuntu 22.04 which is currently not supported by Metal.

Any idea how our team can solve this issue? @davorchap @jvasilje @TT-billteng @ayerofieiev-tt

**Comment by ayerofieiev-tt** [comment]
> 1. As Metal moved to use clang and libc++ library with C++20 standard, it killed our external project because ours were all libstdc++ with C++17 standard based.
> 2. Therefore, we moved our external project to g++-11 which support C++20 standard. Also, we built Metal with g++-11 and libstdc++. It worked for a short period of time before [this](https://github.com/tenstorrent/tt-metal/commit/a8e9e22476c96d60b86f284f2f4eb83760805968) and [this](https://github.com/tenstorrent/tt-metal/commit/b5e0888819eb18fa15a8f639532bfcf662276b5d).
> 3. The above two commits require g++-12 or above because the reflect library those commits are using does not support g++-11. To get g++-12 or above, we should use Ubuntu 22.04 which is currently not supported by Metal.
>    Of course, we can try g++-12 in Ubuntu 20.04 with source code build but I think it will be very cumbersome.
> 
> Any idea how our team can solve this issue? @davorchap @jvasilje @TT-billteng @ayerofieiev-tt

My recommendation is to explore switching to clang. You have full control over it and it our toolsets, which is good because our current CI will keep things sane. I also understand that sometimes toolset switch can be non trivial. 

Team works on adding support for 22.04, I think it is basically there, but there is some work required on CI side. @TT-billteng @ttmchiou know everything about it.

**Comment by razorback3** [comment]
> My recommendation is to explore switching to clang. You have full control over that. It would also align our toolsets, which is great because our current CI will keep things sane. Keep in mind our CI does not check gcc builds today. But I also understand that sometimes toolset switch can be non trivial. I don’t have full context and so this is the best path forward i see.
> 
> Team works on adding support for 22.04. There is some work required on CI side and it looks like that is going to take a couple more weeks. @TT-billteng @ttmchiou know everything about it.

Actually, it is not a problem of clang itself but c++ headers and libraries it uses when _-stdlib=libc++_ compile option is given. And this standard mismatch causes linking issues when mixed with other various libraries.
Anyway, looking forward to the answers from Bill and Michael.
Thanks.

**Comment by ayerofieiev-tt** [comment]
> > My recommendation is to explore switching to clang. You have full control over that. It would also align our toolsets, which is great because our current CI will keep things sane. Keep in mind our CI does not check gcc builds today. But I also understand that sometimes toolset switch can be non trivial. I don’t have full context and so this is the best path forward i see.
> > Team works on adding support for 22.04. There is some work required on CI side and it looks like that is going to take a couple more weeks. @TT-billteng @ttmchiou know everything about it.
> 
> Actually, it is not a problem of clang itself but c++ headers and libraries it uses when _-stdlib=libc++_ compile option (rather than stdlibc++) is given. And this standard mismatch causes linking issues when mixed with other various libraries. Anyway, looking forward to the answers from Bill and Michael. Thanks.

You are right!
Sorry, what I meant is to align the toolset to clang with libc++ across the whole codebase. 

I assume that it was considered and is not feasible (or not desirable) at the moment, is that right?

**Comment by razorback3** [comment]
> You are right!
> Sorry, what I meant is to align the toolset to clang with libc++ across the whole codebase. 
> 
> I assume that it was considered and is not feasible (or not desirable) at the moment, is that right?

Yup.
Since we have two important training milestones that both end within 1-2 months, we were looking for a way that needs minimum changes.

Also, as a quick solution, we tried building Metal with clang-17 but with stdc++. However, we encountered a bunch of compilation errors, and we didn't go more.

**Comment by TT-billteng** [comment]
> > You are right!
> > Sorry, what I meant is to align the toolset to clang with libc++ across the whole codebase.
> > I assume that it was considered and is not feasible (or not desirable) at the moment, is that right?
> 
> Yup. Since we have two important training milestones that both end within 1-2 months, we were looking for a way that needs minimum changes.
> 
> Also, as a quick solution, we tried building Metal with clang-17 but with stdc++. However, we encountered a bunch of compilation errors, and we didn't go more.

Are you able to build metal with clang-17 + libc++, build your code with clang-17 + libstdc++, and link them together?

**Comment by TT-billteng** [comment]
Worst case, we can revert the reflect changes until we move to 22.04.

**Comment by razorback3** [comment]
> Are you able to build metal with clang-17 + libc++, build your code with clang-17 + libstdc++, and link them together?

We tried that too but we met lots of undefined reference error.
I think we should add symbol exporting to make it work between two different standards.

**Comment by razorback3** [comment]
> Worst case, we can revert the reflect changes until we move to 22.04.

I think that can be one option.
Since Metal is an open-source project, I hope your team to please consider achieving consensus and pre-announcing on cataclysmic changes such as toolset change.
Thanks!

**Comment by razorback3** [comment]
> This should be discussed in the next week sync with @arakhmati @cfjchu @eyonland

Thanks for the update, Bill.
Also, I'm gonna propose having discussion about this (compiler toolchain, headers for external projects, etc.) in the upcoming f2f meeting (mid of July) in San Jose.

**Comment by eyonland** [comment]
@razorback3 , if at all possible I would prefer not to revert the reflect library.   There is work planned to significantly help simplify building ops that we have been wanting to do for a long time.   
Also, are there changes that you are blocked on to meet your milestone?  What if you just had a fork to depend on until the milestone is reached and then we delete the fork?   Just trying to explore options.

**Comment by TT-billteng** [comment]
I remember we had a conversation at one point that Moreh was planning to ship conda packages. If so, do you use conda internally for development? You can fetch gcc-12 via conda.

**Comment by razorback3** [comment]
> I remember we had a conversation at one point that Moreh was planning to ship conda packages. If so, do you use conda internally for development? You can fetch gcc-12 via conda.

@TT-billteng, can you come to today's weekly TT-Moreh zoom meeting?
Why don't we talk about this at there?
(cc. @adifrancescoTT, @sangwon-chae)

**Comment by razorback3** [comment]
**Last status:** Tried conda gcc-12.
Conda-installed gcc does not include or link system-installed headers (/usr/include) and libraries (/usr/lib).
Since we require many system-installed libraries such as numactl and hwloc, I tried including the system header directory.
Then, it incurs a lot of compilation errors when building Metal.
I guess that this is because the system include directory contains lower versions of c++ headers.

**Comment by yan-zaretskiy** [comment]
@razorback3 I am working on an alternative solution of using gcc-12 from Homebrew. I ran into similar issues to what you describe above, but I was able to overcome them, so that all the dependency libraries and header files are visible by the compiler. I'm now trying to fix the compilation errors. gcc-12's static analyzer detects some new issues that have to be fixed that are not in this branch, I'm afraid.

**Comment by yan-zaretskiy** [comment]
@razorback3 I am working on an alternative solution of using gcc-12 from Homebrew. I ran into similar issues to what you describe above, but I was able to overcome them, so that all the dependency libraries and header files are visible by the compiler. I'm now trying to fix the compilation errors. gcc-12's static analyzer detects some new issues that have to be fixed that are not in this branch, I'm afraid.

**Comment by razorback3** [comment]
> @razorback3 I am working on an alternative solution of using gcc-12 from Homebrew. I ran into similar issues to what you describe above, but I was able to overcome them, so that all the dependency libraries and header files are visible by the compiler. I'm now trying to fix the compilation errors. gcc-12's static analyzer detects some new issues that have to be fixed that are not in this branch, I'm afraid.

Great. Please keep me updated.

**Comment by yan-zaretskiy** [comment]
I was able to compile `tt-metal` with gcc-12. Now I need to update our cmake files to work with both compiler families.

**Comment by yan-zaretskiy** [comment]
I pushed a new PR that updates our code base to work with both clang 17 and gcc 12. It still uses clang 17 by default, but now you can override the default using the standard CMake variables `CMAKE_C_COMPILER` and `CMAKE_CXX_COMPILER` when calling `cmake`.

**Comment by razorback3** [comment]
> I pushed a new PR (#9943) that updates our code base to work with both clang 17 and gcc 12. It still uses clang 17 by default, but now you can override the default using the standard CMake variables `CMAKE_C_COMPILER` and `CMAKE_CXX_COMPILER` when calling `cmake`.

I'll check it within a few hours and leave a feedback.
Thanks!

**Comment by yan-zaretskiy** [comment]
It's a holiday tomorrow here, so I won't be able to respond frequently. In the interest of time, I wanted to write down the steps I took to make this work with Homebrew on Ubuntu 20.04:

```bash
# install homebrew
NONINTERACTIVE=1 /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"

# add brew to your path
(echo; echo 'eval "$(/home/linuxbrew/.linuxbrew/bin/brew shellenv)"') >> /home/ubuntu/.bashrc
source ~/.bashrc

# install the necessary packages
brew install gcc@12 numactl hwloc binutils python@3.8

# Homebrew does not symlink its version of binutils, but we need to use the GNU assembler from it.
# You gotta put it anywhere in your PATH, for example:
sudo ln -s /home/linuxbrew/.linuxbrew/opt/binutils/bin/as /usr/local/bin/as

export CMAKE_PREFIX_PATH=$(brew --prefix)
cmake -B build -S . -DCMAKE_CXX_COMPILER=g++-12 -DCMAKE_C_COMPILER=gcc-12 -G Ninja
```

**Comment by razorback3** [comment]
```
jaehoon@ttdev01:tt-metal $ cmake -B build -S . -DCMAKE_CXX_COMPILER=g++-12 -DCMAKE_C_COMPILER=gcc-12 -G Ninja
-- Using specifed C++ compiler: g++-12
-- Using specifed C compiler: gcc-12
-- The CXX compiler identification is GNU 12.4.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - failed
-- Check for working CXX compiler: /home/linuxbrew/.linuxbrew/bin/g++-12
-- Check for working CXX compiler: /home/linuxbrew/.linuxbrew/bin/g++-12 - broken
CMake Error at /usr/local/share/cmake-3.28/Modules/CMakeTestCXXCompiler.cmake:60 (message):
  The C++ compiler

    "/home/linuxbrew/.linuxbrew/bin/g++-12"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: '/home/jaehoon/tt-metal/build/CMakeFiles/CMakeScratch/TryCompile-M5idnP'
    
    Run Build Command(s): /usr/bin/ninja -v cmTC_33f25
    [1/2] /home/linuxbrew/.linuxbrew/bin/g++-12    -o CMakeFiles/cmTC_33f25.dir/testCXXCompiler.cxx.o -c /home/jaehoon/tt-metal/build/CMakeFiles/CMakeScratch/TryCompile-M5idnP/testCXXCompiler.cxx
    [2/2] : && /home/linuxbrew/.linuxbrew/bin/g++-12   CMakeFiles/cmTC_33f25.dir/testCXXCompiler.cxx.o -o cmTC_33f25   && :
    FAILED: cmTC_33f25 
    : && /home/linuxbrew/.linuxbrew/bin/g++-12   CMakeFiles/cmTC_33f25.dir/testCXXCompiler.cxx.o -o cmTC_33f25   && :
    /usr/bin/ld: /home/linuxbrew/.linuxbrew/Cellar/gcc@12/12.4.0/bin/../libexec/gcc/x86_64-pc-linux-gnu/12/liblto_plugin.so: error loading plugin: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.33' not found (required by /home/linuxbrew/.linuxbrew/Cellar/gcc@12/12.4.0/bin/../libexec/gcc/x86_64-pc-linux-gnu/12/liblto_plugin.so)
    collect2: error: ld returned 1 exit status
    ninja: build stopped: subcommand failed.
    
    

  

  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:21 (project)


-- Configuring incomplete, errors occurred!
```

@yan-zaretskiy, the error message says it require glibc 2.33 which is not found.
Did you install it manually?

My glibc is 2.31 in ubuntu 20.04.
`ldd (Ubuntu GLIBC 2.31-0ubuntu9.16) 2.31`

**Comment by yan-zaretskiy** [comment]
@razorback3 I did not install `glibc` manually, but I do see it installed. For reference:
```
ubuntu@tt-metal-dev-yan-n150-1:~$ brew list
binutils  ca-certificates  gcc@12  glibc  hwloc  isl   libedit  libmpc  libtirpc  libxau  libxcrypt  libxext  linux-headers@5.15  mpdecimal  ncurses  openssl@3   readline  tcl-tk  xorgproto  zlib
bzip2     gcc              gdbm    gmp    icu4c  krb5  libffi   libnsl  libx11    libxcb  libxdmcp   libxml2  lz4                 mpfr       numactl  python@3.8  sqlite    unzip   xz         zstd
```

**Comment by razorback3** [comment]
Output of `brew list --version`:
I think it is same as you.
```
ubuntu@tt-metal-dev-moreh-wh-8:~/tt-metal$ brew list --version
binutils 2.42
bzip2 1.0.8
ca-certificates 2024-07-02
gcc 14.1.0_1
gcc@12 12.4.0
gdbm 1.23
glibc 2.35_1
gmp 6.3.0
hwloc 2.11.0
icu4c 74.2
isl 0.26
krb5 1.21.3
libedit 20240517-3.1
libffi 3.4.6
libmpc 1.3.1
libnsl 2.0.1
libtirpc 1.3.3
libx11 1.8.9
libxau 1.0.11
libxcb 1.17.0
libxcrypt 4.4.36
libxdmcp 1.1.5
libxext 1.3.6
libxml2 2.12.8
linux-headers@5.15 5.15.161
lz4 1.9.4
mpdecimal 4.0.0
mpfr 4.2.1
ncurses 6.5
numactl 2.0.18
openssl@3 3.3.1
python@3.8 3.8.19
readline 8.2.10
sqlite 3.46.0
tcl-tk 8.6.14
unzip 6.0_8
xorgproto 2024.1
xz 5.4.6
zlib 1.3.1
zstd 1.5.6
```

I found that even I try to compile a very simple hello world printf program with g++-12, it fails to compile with same error message I met when trying to build the Metal.
```
ubuntu@tt-metal-dev-moreh-wh-8:~/workspace$ g++-12 test1.cpp 
/usr/bin/ld: /home/linuxbrew/.linuxbrew/Cellar/gcc@12/12.4.0/bin/../libexec/gcc/x86_64-pc-linux-gnu/12/liblto_plugin.so: error loading plugin: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.33' not found (required by /home/linuxbrew/.linuxbrew/Cellar/gcc@12/12.4.0/bin/../libexec/gcc/x86_64-pc-linux-gnu/12/liblto_plugin.so)
collect2: error: ld returned 1 exit status
```

Through the above error message, I suspect that the compiler is linking the host glibc, not the brew-installed glibc.
So I give `-v` to the compiler and found out that LIBRARY_PATH is set as follows:
```
ubuntu@tt-metal-dev-moreh-wh-8:~/workspace$ g++-12 test1.cpp -v
...(omitted)...
LIBRARY_PATH=/home/linuxbrew/.linuxbrew/Cellar/gcc@12/12.4.0/bin/../lib/gcc/12/gcc/x86_64-pc-linux-gnu/12/:/home/linuxbrew/.linuxbrew/Cellar/gcc@12/12.4.0/bin/../lib/gcc/12/gcc/:/home/linuxbrew/.linuxbrew/Cellar/gcc@12/12.4.0/bin/../lib/gcc/12/gcc/x86_64-pc-linux-gnu/12/../../../:/lib/x86_64-linux-gnu/:/lib/:/usr/lib/x86_64-linux-gnu/:/usr/lib/
...(omitted)...
```

It does not point to the directory where brew-installed glibc resides, which is 
`/home/linuxbrew/.linuxbrew/Cellar/glibc/2.35_1/lib/libc.so`

So, Yan, can you share your output from the compiler?

**Comment by yan-zaretskiy** [comment]
Here's the equivalent output on my VM without truncation, just in case: https://pastebin.com/FSdD4Chz

**Comment by razorback3** [comment]
> Here's the equivalent output on my VM without truncation, just in case: https://pastebin.com/FSdD4Chz

I took several other clean ubuntu 20.04 environments and tried the same but couldn't make it.

After many trials and errors I found that I should do this after gcc installation:
`brew link --force binutils`

Didn't you missed it in your manual?

**Comment by yan-zaretskiy** [comment]
> > Here's the equivalent output on my VM without truncation, just in case: https://pastebin.com/FSdD4Chz
> 
> I took several other clean ubuntu 20.04 environments and tried the same but couldn't make it.
> 
> After many trials and errors I found that I should do this after gcc installation: `brew link --force binutils`
> 
> Didn't you miss it in your manual?

I deliberately avoided this step, because it could mess up with the system version of `binutils`. I don't understand why the Homebrew's gcc does not see its own glibc on your systems...

**Comment by yan-zaretskiy** [comment]
Ah, that's because of the linker. I use the LLVM linker, that's why I didn't see this problem. Your solution of force-linking binutils is correct.

**Comment by razorback3** [comment]
Hey, @yan-zaretskiy.
I think Moreh is fully ready to start a new journey with C++20 compilers.
Since you raised the comprising PR (#9943), hope it is merged to the main soon.
I will close this PR after it is merged to the main.
Big thanks for your work 😄

**Pull request closed without merging**
